### ISSUE TITLE
Lazy analysis

### ISSUE THREAD

**Issue by mainej** [clojure-lsp/clojure-lsp#990]
Inspired by [this article](https://rust-analyzer.github.io/blog/2020/07/20/three-architectures-for-responsive-ide.html), I'm researching a new approach to analyzing and linting. I've been calling this approach "lazy analysis".

Lazy analysis, which I'll explain in a moment, has the potential to reduce the amount of analysis that clojure-lsp needs to do, and to limit the number of elements that must be inspected when querying the analysis. I believe this will improve performance at startup and during use, in projects large and small.

The core insight is that when a user is looking at a file, **most clojure-lsp features only need analysis from that file and its immediate neighbors**.

But that's it! Anything further away—transitive dependencies and dependents—won't affect these features.

To be more precise, features that need var _definitions_ (like go-to definition, view docstring, etc.) need to look only in the current namespace and its direct dependencies. Features that need var _usages_ (like show references, rename, incoming call hierarchy, etc) need only the current namespace and its direct dependents.

See the table below for a more complete picture of which features need which files. See the [Cons](#cons) section for features that need broader analysis.

| Features                         | self | dependencies: var def | dependents: var usage |
|----------------------------------|------|-----------------------|-----------------------|
| call hierarchy: incoming         | x    |                       | x                     |
| call hierarchy: outgoing         | x    | x                     |                       |
| code lens references             | x    |                       | x                     |
| completion                       | x    | x                     |                       |
| diagnostics: general             | x    |                       |                       |
| diagnostics: unresolved var      | x    | x                     |                       |
| diagnostics: unused public var   | x    |                       | x                     |
| docs: docstring                  | x    | x                     |                       |
| docs: signature                  | x    | x                     |                       |
| format                           | x    |                       |                       |
| goto: declaration                | x    |                       |                       |
| goto: definition                 | x    | x                     |                       |
| goto: implementations            | x    |                       | x                     |
| goto: references                 | x    |                       | x                     |
| refactor: coll: change-coll      | x    |                       |                       |
| refactor: coll: cycle-coll       | x    |                       |                       |
| refactor: create-test            | x    |                       | ? (test ns)           |
| refactor: cycle-privacy          | x    |                       |                       |
| refactor: drag: drag-backward    | x    |                       |                       |
| refactor: drag: drag-forward     | x    |                       |                       |
| refactor: fn: create-function    | x    |                       |                       |
| refactor: fn: demote-fn          | x    |                       |                       |
| refactor: fn: extract-function   | x    |                       |                       |
| refactor: fn: promote-fn         | x    |                       |                       |
| refactor: inline-symbol          | x    |                       |                       |
| refactor: let: expand-let        | x    |                       |                       |
| refactor: let: introduce-let     | x    |                       |                       |
| refactor: let: move-to-let       | x    |                       |                       |
| refactor: move-form              | x    | x                     | x                     |
| refactor: ns: add-import         |      |                       |                       |
| refactor: ns: add-require        |      | ?                     |                       |
| refactor: ns: clean-ns           | x    |                       |                       |
| refactor: resolve-macro          | x    |                       |                       |
| refactor: sort-map               | x    |                       |                       |
| refactor: thread: first          | x    |                       |                       |
| refactor: thread: first-all      | x    |                       |                       |
| refactor: thread: last           | x    |                       |                       |
| refactor: thread: last-all       | x    |                       |                       |
| refactor: thread: unwind         | x    |                       |                       |
| refactor: thread: unwind all     | x    |                       |                       |
| rename                           | x    |                       | x                     |
| symbols: color (semantic tokens) | x    |                       |                       |
| symbols: highlight usages        | x    |                       |                       |
| symbols: in file                 | x    |                       |                       |
| test tree                        | x    |                       | ? (test ns)           |

This has several implications:

1. Files don't need to be analyzed until they're opened, or one of their neighbors is opened.
2. A `utils` namespace, which many namespaces depend on but which doesn't have any dependencies itself, may only ever need to be analyzed for its var definitions.
3. Only a small fraction of most library files need to be analyzed. Many libraries have a small number of "public" namespaces, backed by several supporting namespaces. Since a project tends to depend only on the public namespaces, the supporting namespaces don't need to be analyzed.
4. When querying the analysis, usually you need to look only at neighboring files.

The first point gives this proposal the "lazy" moniker. It'll be the biggest change. To elaborate briefly, rather than analyzing every file at startup, we'd gather only the namespace definitions and usages. (@borkdude, this is why I've been interested in analysis of the ns forms.) We'd use the namespaces usages to build a dependency graph. Then, we'd wait until the user opened a file. When they did, we'd analyze it and the files one "step" away from it. As the user opened more files, we'd expand the frontier of analyzed files. This slowly expanding analysis frontier is the "lazy" part.

Files beyond the analysis frontier won't need to be analyzed at all. (Often, and especially in large projects, users work on one "corner" of a project at a time.) By being lazy we can postpone or eliminate huge swaths of analysis, particularly at startup.

To elaborate on the second point, it can be 5x faster to analyze a file's var definitions _shallowly_. Therefore, files that are dependencies of opened files but aren't opened themselves don't need a full analysis.

To elaborate on the third point, take clojure-lsp itself as an example. At the time of writing, it has 52 project files and 646 external files (most of them in jars). But of the external files, only 58 are dependencies of a project file (and even some of those are duplicates, when two versions of a lib are on the classpath, which happens surprisingly often.) So, most of the external files never need to be analyzed. (We'd need to teach clj-kondo how to analyze a single entry within a jar, which it cannot do currently.)

To elaborate on the fourth point, consider "go-to definition". Currently, in the worst case, to find the definition of the element under the cursor`clojure-lsp.queries/find-definition` inspects every element in the analysis, across the entire project and all of its deps. But the definition must be in the current file or one of its dependencies. By using the dependency graph, we could focus `clojure-lsp.queries/find-definition` on the dependency files only, a tiny fraction of the data that must be inspected currently. This would curtail an entire class of performance problems that crop up periodically.

To summarize, lazy analysis would A) reduce the total cost of analysis by avoiding analysis of many files, B) amortize the remaining cost of analysis over the life of the text editor session, and C) reduce the work done to interpret the analysis.

## Cons

Along with these pros, lazy analysis comes with several cons.

### Features that need broad analysis

First, there are a few features that need broader analysis. The table below lists them, and discussion of each one follows.

| Features                     | project: ns | project: var def | project: var usage | external: ns | external: var def |
|------------------------------|-------------|------------------|--------------------|--------------|-------------------|
| diagnostics: full proj       | x           | x                | x                  | x            | x                 |
| symbols: all in workspace    |             | x                |                    |              |                   |
| refactor: auto require/refer | x           | x                |                    | x            | x                 |
| refactor: move-form          | x           | ?                | ?                  |              |                   |
| completion: all nses         | x           |                  |                    | x            |                   |
| completion: others?          | ?           | ?                | ?                  | ?            | ?                 |

First, clojure-lsp will report when there's lint (diagnostics) in any file in the whole project. Of course, it can't show diagnostics of files it hasn't analyzed, so lazy analysis will break this feature. I expect some users will consider this a significant loss in functionality, and so I think we have to plan to make lazy analysis optional. (There are other reasons to make it optional—for the CLI, but also so we can rollback if any major issues are discovered.) Other users, especially those that work in very large projects, may be willing to give up this functionality in exchange for better performance. They'll probably still want full-project diagnostics, but perhaps they'll be willing to postpone it to a CLI invocation of `clojure-lsp diagnostics` in a commit hook or on CI.

Second, to show all symbols in the entire workspace, you need at least shallow var definitions of every project file. When a user requests workspace symbols, the lazy analysis implementation could request shallow analysis of all project files, so this feature could still be supported, though it would be slower on first invocation.

Third, if you type `join` and invoke the actions menu, clojure-lsp will offer to automatically add a `:refer [join]` to the ns declaration, with options for which namespace to refer. I'm not sure whether I think this feature is magical or gimmicky. First, `:as` [is preferred](https://guide.clojure.style/#prefer-require-over-use) over `:refer`, so I'm not convinced we should be offering `:refer`s at all. I would prefer if users typed `set/join`, which changes the actions menu to suggest an alias to `clojure.set`, since that only needs ns declaration analysis. Second, some of the namespace suggestions are plausible to me (`clojure.string`, `clojure.set`, `medley.core`, `rewrite-clj.paredit`) but others aren't (`clojure.test.check.rose-tree`, `clojure.core.rrb-vector.fork-join`). These seem pretty obscure, and probably aren't what I'm looking for.

If this auto-refer feature is to be maintained, we need a shallow var definition analysis of every project file _and_ of every dep file. (Though perhaps we could skip it for dep files that aren't neighbors of project files. This would remove the implausible (to me) suggestions like `clojure.test.check.rose-tree`.) However, I expect that we'd sacrifice a lot of startup performance gains. This will have to be researched more, and again, may be a feature that some users are willing to forego for better performance.

Fourth, move-form needs a lot of analysis, at least about the source and destination namespaces, but perhaps also of their neighbors. This is a new feature and I'm not familiar with its details, so I'd need to research it, or have @snoe think about it. Like workspace-symbols, this can probably be made to work with lazy analysis, though it might become slower.

And finally, completion needs ... a lot of analysis. To be honest, I have trouble reasoning about which analysis it needs. There's a chance it will work fine with system-wide ns declaration analysis (gathered while building the dependency graph) and neighbor analysis (gathered when a file is opened). But, it needs research—it may need system wide shallow var definitions. @snoe and @ericdallo, you seem to have the best understanding of that code.

### Blocking on analysis

More things will need to block when a file is first opened, while the analysis of the file and its neighbors runs. This will mean a slower "first paint" of semantic tokens, code lenses, and refactorings (light bulbs), among other things. This is the downside of amortizing analysis over the editor session.

Fortunately, this performance hit will only be at did-open, not did-change, because changes to the file won't need any more re-analysis of neighbors than we currently do. More generally, as the analysis frontier expands, each newly opened file will be less likely to have many unanalyzed neighbors.

### More complicated analysis tracking

An associated con is that the logic for coordinating analysis will become significantly more complicated.

We'll need to know which files have shallow or full analysis and which are in the process of getting more analysis done. Analysis needs to be batched and debounced. And we'll need good ways to block waiting for analysis to be finished.

See [below](#implementation) for my first thoughts on organizing this complexity.

### More complicated linting

Linting is the area I've made the least progress on, and so I'm most worried about it.

There are (at least?) two linters that need cross-file analysis: `unresolved-var` and `clojure-lsp/unused-public-var`.

Conceptually, `unresolved-var` needs to know what vars are defined in its dependencies, to add lint for `my-real-ns/an-undefined-var`. This linter is managed internally by clj-kondo. For it to work correctly, a namespace's dependencies need to be (shallowly) analyzed either before it's analyzed, or at the same time. Currently, first we analyze the dep files shallowly in batches and then the project files fully in one big analysis, satisfying this requirement. To maintain correctness, when a file is opened, lazy analysis will have to trigger (and wait for) shallow analysis of dependencies before starting full analysis of the namespace and its dependents. Coordinating these two types of analysis adds to the complexity.

`clojure-lsp/unused-public-var` is the opposite—it needs analysis of _dependents_ to work correctly. It uses this analysis to know whether any of a file's vars are unused by its dependents.

When a file is changed, the changes can affect the lint for the file itself as well as its neighbors (see table below). All of this will require more coordination, with linting waiting on the neighbor analysis, and triggering re-linting of neighboring files at the right times.

The code for `clojure-lsp/unused-public-var` is already complicated, so we'd be trading one complicated implementation with another.

| file change       | self                                                                                                   | dependency                                             | dependent                       |
|-------------------|--------------------------------------------------------------------------------------------------------|--------------------------------------------------------|---------------------------------|
| add usage         | remove `upv` lint (first usage) <br/> update code lens                                                 | remove `upv` lint (first usage) <br/> update code lens |                                 |
| remove usage      | add `upv` lint (last usage) <br/> update code lens                                                     | add `upv` lint (last usage) <br/> update code lens     |                                 |
| create definition | add `upv` lint (when unused) <br/> remove `unresolved-symbol` lint (when used) <br/> create code lens  |                                                        | remove `unresolved-var` lint    |
| remove definition | remove `upv` lint (when unused) <br/> add `unresolved-symbol` lint (when used) <br/> remove code lens? |                                                        | add `unresolved-var` lint       |
| change arities    |                                                                                                        |                                                        | add/remove `invalid-arity` lint |

### General implementation strategy

As described above, the general strategy would be:

1. At startup analyze only the namespace definitions and usages.
2. Make a dependency graph.
3. Wait until the user opened a file.
4. When they did, analyze it and its neighbors.
5. As the user opened more files, expand the frontier of analyzed files.
6. As files are changed, update the dependency graph and trigger re-analysis as needed.

### Initial dependency graph

Clojure namespaces and their dependencies are laid out in a directed acyclic graph. To construct this graph, we have two options. First we could construct it lazily. At startup, the graph would be empty. We'd wait for did-open of the first file, analyze it, get its neighbors, and then analyze them. This sounds simple, but I think it would introduces other problems with linting.

Alternatively, as proposed above, we could eagerly analyze all the ns declarations (but nothing else) and use them to construct the graph at startup. This would let us kick-off file and neighbor analysis simultaneously when a file is opened. I think this would be faster and have fewer problems with linting.

I experimented a bit with clojure.tools.namespace, and found that the dependency graph for clojure-lsp and all of its deps can be constructed in less than one second. Perhaps we could use clojure.tools.namespace, or perhaps clj-kondo would have similar performance. In any case, I think we can generate the initial dependency graph quickly during startup, even for large projects.

In any case, whenever a user changes the namespace usages we'll have to update the dependency graph and trigger analysis of new dependencies. This has unfortunate consequences for linting, as described above. Needs more thought.

### Overwriting analysis

Usually a file will be a dependent of some files and a dependency of others. If many of those neighbors are opened simultaneously (e.g. at startup), we might trigger both a shallow and a full analysis of that file. We should be careful to avoid overwriting a full analysis with a shallow analysis.

We should also be careful to avoid overwriting a newer analysis with an older analysis, even in the face of debouncing and batched analysis.

### Managing complexity

To tame the complexity of lazy analysis, I've found it productive to think of files and the analysis process as state machines.

(@snoe and @ericdallo, over the last few months we've had several discussions about how to coordinate analysis. I've suggested modeling with priority queues, but I now think state machines better fit the problem.)

To go into more detail, we'd track the analysis state of each file. We'd keep track of where we were on analyzing the file and have rules for when to transition between states.

Using annotation inspired by [XState](https://github.com/statelyai/xstate) and [clj-statecharts](https://lucywang000.github.io/clj-statecharts/) here's a first draft of the file state machine. The state machine encodes the rules for:
* submitting a file to the analysis scheduler
* ensuring a file always progresses from less analyzed to more analyzed
* ensuring that old analysis doesn't overwrite new analysis
* allowing additional analysis to be scheduled even if some analysis is already queued or in progress.

```
file-analysis-state
  unanalyzed {
    on {
      dependent-opened: shallow
      dependency-opened: full
      file-opened: full
      file-changed: full
    }
  }
  shallow {
    initial: queued
    ;; TODO: when leaving this state, also remove from shallow analyzer queue?
    on {
      dependency-opened: full
      file-opened: full
      file-changed: full
    }
    states {
      queued {
        actions: [
          save-shallow-version
          shallow-scheduler/enqueue
        ]
        on {
          shallow-started: in-progress
        }
      }
      in-progress {
        on {
          shallow-finished: idle, guard latest-shallow-version
        }
      }
      idle { }
    }
  }
  full {
    initial: queued
    on {
      file-changed: queued (drops in-progress analysis)
    }
    states {
      queued {
        actions [
          save-full-version
          full-scheduler/enqueue
        ]
        on {
          full-started: in-progress
        }
      }
      in-progress {
        on {
          full-finished: idle, guard latest-full-version
        }
      }
      idle { }
    }
  }
```

Similarly, I think it's useful to think of the analysis scheduler as a state machine, or rather as two state machines, one for shallow analysis and one for full. I'm less confident this is well modeled, but as a first draft this state machine encodes the rules for:
* batching analysis, when several files are submitted at the same time
* debouncing analysis, when several did-change events happen in quick succession
* re-enqueueing files even if analysis is already in progress
  
```
;; TODO: this is modeled as two schedulers. Should it be one? Or one per file? Or...
;; NOTE: queued/running sets should store version:
;;   1. to avoid queueing older version
;;   2. to report which version was analyzed when file-analysis/shallow-finished (or full)
;;      is triggered on file statemachine
analysis-scheduler-state {
  paused {
    initial: idle
    states {
      idle {
        on {
          queued {
            actions: add-to-queued-set
            to: debouncing
          }
        }
      }
      debouncing {
        delay 300ms: running
        on {
          queued {
            actions: add-to-queued-set
            to: running (when batch-size-reached) ;; start batch, even before debounce completes
            to: debouncing ;; restart delay
          }
        }
      }
    }
  }
  running {
    actions [
      move-from-queued-to-running-set
      dispatch-analysis-to-kondo
      file-analysis/shallow-started (or full) for each file
    ]
    on {
      queued {
        actions: add-to-queued-set
      }
      analyzed {
        actions [
          remove-from-running-set
          file-analysis/shallow-finished (or full) for each file
        ]
        to: running (when seq queued) ;; assumes queued has waited awhile, meriting immediate run
        to: paused
      }
    }
  }
}
```

As mentioned, I haven't come up with a clear plan for coordinating linting. It may involve a state machine with the following per-file states.

```
lint-state
  unlinted
  waiting-for-neighbors
  linting
  linted
```

### Blocking on state

As discussed, many things—lazy analysis itself included—will need to query the analysis state of files and wait for analysis to reach a particular state. I'm not sure how to implement this. Polling would probably work. Perhaps it can be achieved with `add-watch` on whatever holds the file state.

### Order of implementation

There are different ways to order this work. For example, before fully implementing lazy analysis, we could start by maintaining a dependency graph. With that information, we could change `clojure-lsp.queries` to inspect only neighboring files. Perhaps we could restrict the analysis done during an uncached startup to the neighbors of project files.

Are there other ways to split this project into smaller pieces?

## And more

There's more to do though. I haven't worked out how to coordinate everything. Something needs to listen to did-open and did-change events, update the dependency graph, submit newly discovered dependencies, and correctly sequence linting.

I think several threads can receive a did-change simultaneously. To ensure nothing gets lost, we need to make state machine updates atomic. How to do this? Refs? Agents? An async channel?

### Types and amount of analysis

I've tried to present one way of doing the analysis, but there are actually several approaches.

* Shallow analysis helped reduce the amount of time analyzing external files by 60-70%. But skipping non-neighboring external files would also cut that time. (In clojure-lsp, we could skip 90% of the external files.) Of course it would be nice to do both, but perhaps having two types of analysis adds too much complexity. Would it be better to always do full analysis? Or full analysis of internal files and shallow analysis of external files, as we do currently?
* Would it be too slow to do shallow analysis of project files and their neighbors eagerly at startup? Would that simplify linting? Probably "yes", it would simplify unresolved-var, but "no" not clojure-lsp/unused-public-var.
* With this lazy approach, there's less of a boundary between internal and external files. Should we start doing full analysis (including linting) on external files? Or do we already do that via `run-kondo-on-text!`?

### Miscellaneous questions

* What special handling do cljc and cljs files need? 
  * In particular, do we need to be careful about how we interpret ns declarations for different file types? Does clojure-lsp do anything to handle this currently? That is, if you're in a CLJC file and a single alias points to two files, one CLJ and one CLJS, how does go-to-definition decide which file to go to?
* Will this affect stubs and java analysis?
* One thing we can partially ignore is that the file-to-namespace relationship is many-to-many... one file can define many namespaces and one namespace can be defined in several files. Fortunately, this is rare outside of "scratch" files, and so it's probably safe to assume that there's a one-to-one relationship. However, surprisingly often the classpath includes several versions of a library, probably when several different tools contribute to the classpath. This means some namespaces are defined in several files. clojure-lsp/clj-kondo don't really do anything to handle this currently, but perhaps it needs to be considered.
* Currently we store a cache of external analysis on disk, for faster startup. As external files are lazily analyzed, should they be added to the cache? (I.e. make the update of that cache ongoing, instead of once.) Or do we remove the cache, because external analysis will be more limited? Or something else?
* The tables are based on my intuitive understanding of features and the files they need, not on code. What have I missed?

## Summary

I think there's a lot of potential here. But clearly I have a long way to go before this can be implemented. As I flesh out various parts, I'll update this issue.

I don't think I have any specific requests of others at the moment. I'd like it if clj-kondo knew how to analyze the first ns declaration in a file. I'd also like it if it knew how to analyze a single entry within a jar. But I think this project has to prove the necessity of those features first.

Otherwise, if anyone has feedback, I'd appreciate it. Is this a worthwhile direction to take clojure-lsp? Are the pros worth the cons? Can we work around the cons, and if so, do you have advice on how?

**Comment by borkdude** [comment]
> I'd like it if clj-kondo knew how to analyze the first ns declaration in a file.

Actually I have some use for this myself. When linting a directory or jar file (when linting in parallel these are "groups" that are linted in one thread), I'd like to sort the sources on namespace topological order and then lint them in that order. This will ensure certain things to work better, like `ns-analysis` which requests the analysis of other another namespace.

**Comment by ericdallo** [comment]
Amazing deep dive @mainej, really glad you took the time to check each feature and how that would behave, thanks for that.

I like lazy analysis as AFAIK is the way most famous/used LSP servers do, like Dart and Rust, certainly would make clojure-lsp work faster on any Clojure project, although I'm surprised how far we got with full analysis, the recent changes we did really improved clojure-lsp in a way we didn't receive any performance complains recently, even so, I'd be happy to see lazy analysis working someday.

I think the biggest concern I have is for sure complexity, you presented excellent points which I agree with, and to be honest I feel a little bit lost about how we could kick off this 😅. I think we could introduce some kind of flag and evolving since ns dependency graph until we have a fully functional feature, again, it makes me wonder how many ifs/branches we would need to introduce and the complexity.

### About cons

For sure, diagnostics is the most affected feature here, followed by completion. I think most users don't use these completion features that rely on whole analysis that much, but we can't just ignore it too, needs more thought indeed.
On the other hand, maybe we could workaround diagnostics, and do a post-startup async full project findings analysis **only** on kondo? if that would be faster than bringing analysis which I think it would.

> ### Blocking on analysis

Yeah, I'm curious to know how much time would be needed to open a namespace with dozens of requires, but probably would be significant for those ns with hundreds and hundreds of lines (which usually also have multiple requires).

> ### Order of implementation

> Are there other ways to split this project into smaller pieces?

I know there is a lot to discuss to have this feature, so maybe we could try to split into parts indeed, like introducing a flag, ns dependency graph, analysis using this feature etc.

> ### Miscellaneous questions

> , how does go-to-definition decide which file to go to?

It's actually by luck of analysis sort, pretty bad, but it's something is not that common, even so we should fix that somehow.

> Will this affect stubs and java analysis?

I don't think so, we do a scan of java analysis only with kondo and it's pretty isolated and fast enough

>  This means some namespaces are defined in several files. clojure-lsp/clj-kondo don't really do anything to handle this currently, but perhaps it needs to be considered.

Yes, needs more though too, I can't quickly think on a easy fix for that

> Currently we store a cache of external analysis on disk, for faster startup. As external files are lazily analyzed, should they be added to the cache?

Good question, to be honest, I'd leave this to the future after we have a more concise lazy analysis feature.

> The tables are based on my intuitive understanding of features and the files they need, not on code. What have I missed?

I think you mentioned all features I had in mind, especially diagnostics and workspace symbols

**Comment by mainej** [comment]
@ericdallo and @borkdude, thanks for reading this over and clarifying how things work! That'll help me make sure things keep working.

> split into parts, like introducing a flag, ns dependency graph, analysis using this feature etc.

I've been thinking the same thing. I'll start with maintaining a dependency graph, even if it's unused. When we're confident that graph is reliable, I'll switch analysis to use it. Later, I'll work on lazy analysis, though I'll keep thinking about it in the meantime.

> introduce some kind of flag

I can envision two flags—one for dependency-graph based queries, and one for lazy analysis. 

IMO, the dependency-graph based queries flag should be temporary, since that should work whether you're using a dependency graph built with eager analysis or with lazy analysis. 

But at first a flag will be useful. As clojure-lsp developers we can use it and maybe ask the community to test it as an "experimental" feature. After we're confident it works, we can remove the code paths that use the old way.

> makes me wonder how many ifs/branches we would need to introduce

For dependency-graph based queries, the problem is manageable. We can isolate all the branches to a single namespace.

```clojure
(ns clojure-lsp.dep-graph)

(defn use-dep-graph? [db]
  (settings/get db [:experimental :dep-graph-queries] false))

(defn self-and-dependent-analysis [db file]
  (if (use-dep-graph? db)
    (select-keys (:analysis db) (self-and-dependent-files db file))
    ;; TODO: remove this code path when dep graph is out of experimental status
    (filter-project-analysis db)))

;; elsewhere

(queries/find-references
 (dep-graph/self-and-dependent-analysis db (:filename element))
 element)
```

Managing the branching for lazy analysis could be trickier. There are lots of places that will need to know what kind of analysis is being done. One possible way to handle that is with the "replace conditional with polymorphism" refactoring. We'd make an `Analyzer` protocol with two implementations, an `EagerAnalyzer` and a `LazyAnalyzer`, and assoc the proper one into the `:components` at init. Then "all" you'd have to do is figure out what the protocol needs to look like and move the analyzer-type specific code to the implementations. Undoubtedly that's easier said than done.

Anyway, I'll start with the dependency graph and the queries. I'll plan to open separate issues/PRS for that work. I'll continue to add to this Issue as I make progress on other lazy analysis questions.

**Comment by ericdallo** [comment]
> But at first a flag will be useful. As clojure-lsp developers we can use it and maybe ask the community to test it as an "experimental" feature. After we're confident it works, we can remove the code paths that use the old way.

Totally agree!

> For dependency-graph based queries, the problem is manageable. We can isolate all the branches to a single namespace.

Good pseudocode! looks promising to me.

> Undoubtedly that's easier said than done.

For sure, but sounds like a good approach compared with others I thought

**Comment by mainej** [comment]
I've been thinking about linting, and in particular, how to lint references (or in my new terminology, neighbors). This question has been bothering me because when we run analysis, we'll learn of changes to var usages. With this new information, we'll need to re-run unused-public-var linting in neighbors. But to do that, we need analysis of their neighbors. I'm worried that we could end up in a deadlock where analysis can't finish because it's waiting on linting, which is waiting on further analysis.

I still have lots of open questions, but here's the progress I've made.

First, it's useful to understand the existing `notify-references` algorithm. When a file is changed, we check whether the change added, removed or updated any var usages or definitions. The possible changes are in the table above (add/increment usage, remove/decrement usage, add definition, remove definition, change arities of definition). Then we figure out which neighbor files would have been affected by those changes, lump them together (both dependents and dependencies), and ask for them to be re-analyzed and re-linted. We also tell the editor it needs to update code lenses. To keep the re-analysis efficient, we do it in one big batch—first, asking clj-kondo to process all the affected files in parallel, and second, linting unused-public-var in a single pass of the analysis, instead of one pass per file.

Looking at the table above and comparing it to this algorithm, we can notice some inefficiencies.

First, though we ask for the neighbor files to be re-analyzed and re-linted, we don't actually need them to be analyzed. They haven't changed, so their analysis will be the same. That is, we could ask for linting but skip analysis. (@ericdallo please correct me if you think this is wrong.)

Second, we treat dependency and dependent files the same, but really they have very different needs.

Suppose the file change removes a usage of a var that's defined in a dependency file. We need to re-run the `clojure-lsp/unused-public-var` lint for that dependency file, but that linting happens entirely within clojure-lsp, and so technically we could bypass clj-kondo. We don't expect the clj-kondo managed linting of the dependency to change.

Alternatively suppose the file change removes a var definition, a definition that was being used in a dependent file. This is the opposite situation. We need to run the clj-kondo linting on the dependent file, to add `unresolved-var`, but the `clojure-lsp/unused-public-var` linting for the dependent file won't change.

And finally, if we don't have any var usage changes, there won't be any code lens changes, and so (maybe) we could skip the code lens update.

All of this points to the conclusion that we could be more efficient by splitting `notify-references` into `notify-dependencies` and `notify-dependents`.

Coming back to lazy analysis, what does all this mean?

When a file changes we need to:
1. Detect new dependencies and if there are any, ensure they have shallow (var definition) analysis
2. Run analysis on the changed file, along with clj-kondo lint (but not yet unused-public-var lint)
3. Notify the file's (open) dependents of changes in definitions
  a. Enqueue clj-kondo linting of these files
4. Notify its (open) dependencies of changes in var usages
5. For the file itself and its affected dependencies, wait for full (var usage) analysis of dependent files
6. Run unused-public-var lint as soon as var usage data is available

And looking at this, we can see we can have **two separate processes**, one that handles steps 1-4 and another that handles 5 and 6.

As files are opened and as the dependency graph changes, we'll have one process that triggers analysis. From that analysis we may discover we need to re-run clj-kondo linting in dependent files. The analysis process can immediately continue with that work. From the analysis we may discover also that we need to run unused-public-var linting in a changed file or in its dependencies, but crucially, the analysis process doesn't need to block on the unused-public-var linting. We'll always be making forward progress on the analysis, moving toward an idle state where we've done all the analysis we need to do.

Alongside that we'll have an unused-public-var linting process. It'll be triggered by changes in var usages and definitions, and it may need to block waiting for dependent file analysis to finish. But, since the analysis process will be working toward an idle state, the unused-public-var linting will eventually unblock. 

Thinking about analysis as a separate process from unused-public-var linting makes me more confident they can both be implemented without worry that we'll end up in a deadlock where linting is waiting on analysis, which is waiting on linting.

I've kind of glossed over clj-kondo linting (i.e. all linting except unused-public-var). This needs some more thought but I think it'll be OK. That is, I think we'll end up with the right lint from clj-kondo if we A) run clj-kondo linting along with analysis on did-change (as we do currently) and B) notify dependent files when var definitions change, and re-run clj-kondo lint for them too. (We'll have to be careful to always merge clj-kondo lint with clojure-lsp lint, instead of overwriting one with the other.)

Will this strategy be performant? Again, setting aside full-project diagnostics, I think we'll be OK. Counting usages should be much faster with dependency-graph based queries. Similarly, it'll be much quicker to find neighbor files in `notify-references`. But, we were getting some performance gains by linting all the neighbor files in one batch. If we make unused-public-var linting a separate process, we could easily lose that batching. So, we'll have to keep an eye on it. If it looks slow, we'll need to re-introduce batching, or perhaps keep caches of usage counts (maybe taking advantage of clj-kondo's `ns-analysis`, which includes cached usage data).

Finally, what about full-project diagnostics? I suspect that dependency-graph queries will help, but that the performance of full-project unused-public-var will still be highly dependent on batching. We'll want to get all the var usages in each file in a single pass over that file, not one pass per dependency. That is, we want an algorithm that is proportional to the number of nodes in the graph, not the number of edges. So, we may need to delegate to the existing algorithm.

Anyway, there are still details to work out, but at least I'm getting a clearer picture of the problem.

**Comment by ericdallo** [comment]
> First, though we ask for the neighbor files to be re-analyzed and re-linted, we don't actually need them to be analyzed. They haven't changed, so their analysis will be the same. That is, we could ask for linting but skip analysis. (@ericdallo please correct me if you think this is wrong.)

Correct, we care for kondo findings not analysis in that moment, analysis should not change for references indeed.

> , and so technically we could bypass clj-kondo. We don't expect the clj-kondo managed linting of the dependency to change.

Correct, but ATM we are using clj-kondo custom-lint-fn feature, but maybe we could manage to re-lint unused-public-vars without calling kondo with what we have on db already

> All of this points to the conclusion that we could be more efficient by splitting notify-references into notify-dependencies and notify-dependents.

Yes, I agree we can improve that, I just wonder how "bad" is that ATM, is that a real problem making those unnecessary scans? how relevant is the performance lost in those cases?

> Thinking about analysis as a separate process from unused-public-var linting makes me more confident they can both be implemented without worry that we'll end up in a deadlock where linting is waiting on analysis, which is waiting on linting.

Yeah, sounds like will help make code clearer as well

> Anyway, there are still details to work out, but at least I'm getting a clearer picture of the problem.

Thanks for sharing, it's making sense to me until now, all concerns sound valid

**Comment by snoe** [comment]
Catching up on this. I admit I may have missed some important details. I think to go forward we'd need to be very specific about the "contracts" we want to have with the user. Probably something we should have generally as a project, but from my understanding is going to be important for this work.

### implications

> 1. Files don't need to be analyzed until they're opened, or one of their neighbors is opened.

> 
> 4. When querying the analysis, usually you need to look only at neighboring files.

Is this correct? `defmethod` and re-frame `reg-*` calls are often required in a `main` file while only the `defmulti` location is required at call sites, if you just open the call site nothing will lead you back to `main` and then to the `defmethod`. Potemkin is another library that also needs to be considered as you cannot just look at ns forms, people also make their own potemkin-like macros that could introduce `(require x) (x/y)` calls, so kondo analysis is important and might mean that deep analysis is necessary. At metabase we have code that loops over all known namespaces on the classloader at runtime looking for defmethods to automatically require them. Tests are also famously never required. namespaces may have direct calls to fully-qualified namespaces without ever using `require` as they know it's loaded by `main`. Unfortunately, in my experience, code that affects global var state on load is much more common than we admit in clojure.

> 3. Only a small fraction of most library files need to be analyzed. Many libraries have a small number of "public" namespaces, backed by several supporting namespaces. Since a project tends to depend only on the public namespaces, the supporting namespaces don't need to be analyzed.

Jumping into a jar and being able to go to definition of supporting vars, functions and namespaces is important. It's unclear if this was considered.

### other comments

> Features that need var usages (like show references, rename, incoming call hierarchy, etc) need only the current namespace and its direct dependents.

I believe this is not true because of 1. above.

> Fourth, move-form needs a lot of analysis, at least about the source and destination namespaces, but perhaps also of their neighbors. This is a new feature and I'm not familiar with its details, so I'd need to research it, or have @snoe think about it. Like workspace-symbols, this can probably be made to work with lazy analysis, though it might become slower.

This probably works, as you move-form from the definition. 

> And finally, completion needs ... a lot of analysis. To be honest, I have trouble reasoning about which analysis it needs. There's a chance it will work fine with system-wide ns declaration analysis (gathered while building the dependency graph) and neighbor analysis (gathered when a file is opened).  But, it needs research—it may need system wide shallow var definitions. @snoe and @ericdallo, you seem to have the best understanding of that code.

I don't see how this works just through dependency graphs of open files.

Consider a new file in a new session. If a user types in `m/map-` they would expect `medley.core/map-vals` and `medley.core/map-keys` (with associated `:require` text-edits) to be completion suggestions. That requires to know that some project namespace requires medley `:as m` as well as what vars exist in the library.

Consider the above if the user just added medley to their dependencies. We don't know that medley.core needed to be analyzed as it's never been required before.

### my concerns

A big thing that I think we've gone away from is picking up changes from the filesystem. It is very important that if you `git pull` new files that usages and lints are updated in the editor.

Along with new files appearing from git, if I pull and the file I have open changed there are things that need to happen. New functions in namespaces may have appeared `(u/new-fn)` and, I believe currently, will be marked `unknown-var` until you re-open the source namespace. So filesystem changes need to also trigger analysis in files that are depended on.

Rename, and find-usages (and other usage based information and operations) must be certain they block while parts of the code are still being analyzed (including after a git pull).  If there are unanalyzed usages we break their code.  I strongly believe we should consider refactors that loop over "usages" and showing "usages" (implementations, tree calls, whatever) that the user can loop over to change manually in the same class of guarantee: **that we presented or operated on  a complete picture of the project**.

Think about `defmulti` a lot for these as I think it is emblematic of the spooky action at a distance parts of clojure code loading. 

Also think about potemkin and making sure that refactors will follow through its macros and other macros that do similar var manipulation.

Lints should not pop in or out long after you opened a file. We were in that state at one point and it was awful. People make edit decisions based on unknown-var or unused-public-var and both false postitives or negatives can really mess up workflow.

Overall, this might still be possible but I'm sort of left wondering why? To me the value of clojure-lsp lies in that whole project understanding, I worked on refactor-clj before this and the incomplete picture of the project that repls have was the main motivating factor for starting this project. Doing a rename and it not finding the unloaded test calls, or the unloaded `user.clj` calls was infuriating. I'm worried this approach leads to a less complete understanding of the project than the repl has. I am willing to trade startup time for correctness and responsiveness while editing any time.

**Comment by mainej** [comment]
Thanks for adding a critical eye @snoe—it's really helpful.

First, a quick aside. I think you're aware of this, but I feel I should clarify... When I describe "shallow analysis" I usually say it means we analyze var definitions only. But that's not quite right. What it really means is that when the analysis encounters a `def`, `defn`, or several other things that define vars, it skips children, rather than recursing into them. But, top-level stuff gets analyzed, including re-frame registrations, potemkin `import-vars`, etc.

Moving on to your criticisms... I think you're primarily pointing to a family of flaws in my reasoning.

The first member of this family can be stated succinctly as: to see var usages from another usage, the queries ([here](https://github.com/clojure-lsp/clojure-lsp/blob/8517f6a54dbb3272b529bf914b409134ad67dc7f/lib/src/clojure_lsp/queries.clj#L350-L353)) A) go to the var definition, then B) go to the var's usages. That is, we look in one of the open file's dependencies and then in _that_ file's dependents. That's _two_ steps away from the open file, not one.

This is true of all "var usages from another usage". In fact, with this way of thinking, `defmulti` isn't a special case. Suppose you're on a usage of a var defined by `defmulti`. There's a single file that defines that var. All of the var's usages and implementations (including both a `main` namespace, and all the `defmethod`s) are in dependents of that defining file. That is, they're at most 2 steps away from the usage you started from, steps that are easily deduced from the `ns` forms. In particular, the usages and implementations aren't anywhere "spooky", though see more about spookiness below.

The more problematic member of the family is keyword usages. In general, keywords don't have a single place where they're defined (though re-frame and core.async "register" keywords, an action which is treated much like a var definition.) Because of this, when finding usages of a keyword, the only option is to look at the entire project. For example if you're asking, "Is this usage of `:bucket` the same as that one over there?" the answer is, "Maybe 🤷."  The ambiguity is an inherent (and IMO immensely useful) part of Clojure. clojure-lsp can't resolve the ambiguity and so it does the most useful thing it can: shows you every usage of `:bucket` anywhere, including places you, as a human, can see are **obviously** not the same.

Assuming I have all of that right, we're on the same page so far. I think I could work around the var usages problem, by having the queries trigger and wait for whatever analysis they need, allowing them to inspect data from files two (or more) steps away from an open file. That'd keep `defmethod`s working. Perhaps you do some crazy stuff to require those `defmethod` files, but I don't think that would matter. Since _they require the `defmulti` file_, they'd be part of the dep-graph anywhere that matters. Similarly for test namespaces; they require the namespace under test, so they'd be part of the dep graph.

But—like full project diagnostics where there could be lint literally anywhere—a keyword usage could be anywhere. The only way to find all the usages of a keyword is to analyze the whole project. And like with full project diagnostics, if this functionality were lost some users might accept it but others would consider it a big enough detriment that they'd want to disable lazy analysis. There are more and more things like that about lazy analysis, and as the number mounts, I'm starting to ask the same question as you—would it be worth it?

With this discussion, I'm feeling less enthusiastic about lazy analysis, but perhaps there are ideas that could be salvaged from it. One of those ideas is dep-graph queries. I still believe they'd have big performance benefits, even without changing the analysis process at all.

Anyway, it'd be helpful to me if we talk about dep graph queries separately, as we discuss pros and cons of the larger proposal. For example: querying usages from another var usage could use a dep-graph (and could be made to work with lazy analysis). Querying usages from another keyword usage couldn't use a dep-graph (or be made to work with lazy analysis), though that wouldn't preclude other queries from using a dep-graph.

Let me address some of your other concerns.

> `potemkin`

Will you clarify your worries about potemkin? I don't think I understand. From what I see, clj-kondo has some special code to treat `import-vars` as a var definition, much like `def`. But clojure-lsp is basically oblivious. The only special handling that clojure-lsp does for potemkin is to make goto-definition navigate from a symbol inside `import-vars` to the original definition, as if `import-vars` created both a var definition and a var usage. (Though [the way clojure-lsp does this](https://github.com/clojure-lsp/clojure-lsp/blob/8517f6a54dbb3272b529bf914b409134ad67dc7f/lib/src/clojure_lsp/queries.clj#L166-L170) looks broken, or at least incredibly fragile. It ignores the namespace of the original var, and just goes to any var definition with the same name. So, if you imported `join`, it could take you randomly to any of the half dozen namespaces that define `join`.)

Does clj-kondo also create a namespace usage from `import-vars`? Not that I see, but even if it does, I still don't see the problem. The dep-graph would notice the namespace-usage, include it, and use it. (If you call `import-vars` inside of another function, that's much like calling `def` within a function, which [is discouraged](https://guide.clojure.style/#dont-def-vars-inside-fns). I'm not saying that people don't do it, but that I don't expect clj-kondo or clojure-lsp to ever automatically handle cases like that. For example, should clj-kondo resolve vars in that namespace? What if that function is never called?)

So yes, potemkin is a little spooky but, thanks to clj-kondo, not in a way that affects my reasoning about dep-graph based queries (or lazy analysis for that matter) as far as I can see.

More generally, yes, you can do spooky things with Clojure. It's a LISP after all. You can have macros that require namespaces at compile time. You can have runtime code that conditionally requires namespaces. Maybe you can even define namespaces at runtime. All of this would make a dep-graph deviate from what's "really going on". But even right now, clojure-lsp doesn't just magically understand what's "really going on". It needs special code in clj-kondo that converts the magic into what clojure-lsp sees as a regular namespace-usage or var-definition. As I'm envisioning things, those would go into the dep graph and the analysis. And as far as I can see, things would keep working. Perhaps you have specific counter examples.

As an example of something that clj-kondo doesn't understand, see the following code.

```
(defn foo []
  (require '[medley.core])
  (medley.core/map-vals identity {}))
```

For me at least, goto-definition fails when I'm on `medley.core/map-vals`. (I also see an `unresolved-namespace` lint.) You could argue that goto-definition should work because the definition of `foo` compiles, doesn't throw any errors when you call `(foo)` and returns the expected result. But, if you wanted that, you'd have to add special code in clj-kondo that understood this form.

The other option, the option clojure-lsp chooses currently, is to say: if you're going to do spooky stuff, that's on you. clj-kondo knows of a few special cases that are common within the community, but if you go outside that realm, we won't help you. In particular, we cannot guarantee that we "operated on a complete view of the project". 

> A big thing that I think we've gone away from is picking up changes from the filesystem. It is very important that if you `git pull` new files that usages and lints are updated in the editor.

I don't disagree, but I believe that's an existing issue (possibly related to the fact that [this code](https://github.com/clojure-lsp/clojure-lsp/blob/8517f6a54dbb3272b529bf914b409134ad67dc7f/lib/src/clojure_lsp/feature/file_management.clj#L222-L226) is commented out and has been since I started working on the project.) Can you clarify how that problem would be made worse by these proposals? That is, if we _were_ handling filesystem changes better, by triggering re-analysis more often, why wouldn't these proposals be able to pick up and use that analysis too?

> Rename, and find-usages (and other usage based information and operations) must be certain they block while parts of the code are still being analyzed

Again, that's true, but isn't how things are done currently, meaning this is already buggy. If anything, having a better model about what analysis is available, is needed, or is currently being done would make it _easier_ to block on analysis.

> Doing a rename and it not finding the unloaded test calls, or the unloaded user.clj calls was infuriating.

I imagine so, but the thing to remember is that though nothing requires test files or user.clj, they're on the classpath and they require other namespaces, which means they'd be part of the dependency graph. 

> Jumping into a jar and being able to go to definition of supporting vars, functions and namespaces is important

Agreed, and it was part of my original considerations. I expected to continue expanding the analysis frontier into dependencies. But if lazy analysis is going to be taken off the table, it won't matter.

> namespaces may have direct calls to fully-qualified namespaces without ever using require as they know it's loaded by `main`

True, that's valid Clojure. But clj-kondo doesn't analyze it that way. It flags it as an unresolved namespace and doesn't create a var usage. This is another case where we don't guarantee we operated on a complete view of the project.

> If a user types in m/map- they would expect medley.core/map-vals and medley.core/map-keys

True, that part of completion wouldn't work with lazy analysis, at least not without blocking on a bunch of additional analysis. Another point against lazy analysis.

Anyway, to summarize my thinking at the moment... lazy analysis has the potential to slash startup time in large projects. At the same time, there are a small handful of reasonably high-profile features that either wouldn't work with lazy analysis, or would introduce complexity and be slower than they are now. If the list were smaller I might be willing to put in more effort arguing for and/or implementing lazy analysis, but at the moment, I'm not feeling motivated.

There are also dep graph queries. Several of the areas of code that either have had performance problems recently or which continue to teeter on the edge of performance problems—completion, notify references, unused-public-var—boil down to the same problem: We can determine a small set of namespaces we're interested in, but the only way to get information about them is to loop through the entire analysis data set, often several times. These areas have the most to gain from dep-graph queries.

This discussion has clarified some nuances of how dep-graph queries would work, and a few cases where they wouldn't be appropriate, but I don't see any new impediments to using them. The plan already was to implement them first, before moving on to lazy analysis. They'll be easier to reason about in the current analysis scheme, where we eagerly run full analysis.

**Comment by snoe** [comment]
@mainej 
>  if we talk about dep graph queries separately,

I think there were three parts to the larger proposal: lazy-analysis, shallow project analysis (that does full analysis closer to the user), and querying using a dep-graph type index. I'm **assuming** from your comments, that you'd like to build the dep-graph from shallow analysis, I'll argue that's not a good idea. 

The idea of a complete dep-graph for query-ability makes perfect sense though.  These are exactly the kind of indices we should look to maintain.

In regards to how we handle things like git pull, that was there to consider if we rewrite the analysis pipeline.

> First, a quick aside. I think you're aware of this, but I feel I should clarify... When I describe "shallow analysis" I usually say it means we analyze var definitions only. But that's not quite right. What it really means is that when the analysis encounters a def, defn, or several other things that define vars, it skips children, rather than recursing into them. But, top-level stuff gets analyzed, including re-frame registrations, potemkin import-vars, etc.

This is helpful, my understanding was different with a more mechanical analysis of the ns form. As long as kondo hooks are triggered and analyzed I think that addresses many of the concerns about the shallowness of the analysis regarding potemkin like macros. 

Yes, there are a number of problems where kondo fails to analyze things that clojure interprets, in my mind those are just part of continually improving the analyzer. For instance, your example is picked up fine for me on master now.
<img width="307" alt="image" src="https://user-images.githubusercontent.com/105012/170768186-643c783c-53c4-4a48-8453-aa497d4bb243.png">

> This is true of all "var usages from another usage". In fact, with this way of thinking, defmulti isn't a special case. Suppose you're on a usage of a var defined by defmulti. There's a single file that defines that var. All of the var's usages and implementations (including both a main namespace, and all the defmethods) are in dependents of that defining file. That is, they're at most 2 steps away from the usage you started from, steps that are easily deduced from the ns forms. In particular, the usages and implementations aren't anywhere "spooky", though see more about spookiness below.

Let me lay out some files. 

```clojure
(ns lib.core)
(defmulti foo ...)

;; ---
(ns example.impl (:require [lib.core :as c]))
(defn go []
  (c/foo ...))

;; ---
(ns example.extension (:require [lib.core :as c]))
(defmethod c/foo ...)

;; ---
(ns example.main
  (:require [example.impl] [example.extension :refer [go]))

(defn main [] (go))

;; ---
(ns example.impl-test (:require [lib.core :as c]))
(defmethod c/foo ...)

;; ---
(ns user)
(comment
  (lib.core/foo ...) ;; called from nrepl that has ns loaded through `example.main`
  )
```

From the usage in `example.impl` a graph "deduced from ns forms" is defined, `[example.impl [lib.core]]` If you open `example.main` you might get a more full view `[example.main [example.extension [lib.core]] [example.impl [lib.core]]]]`. `user` and `example.impl-test` are not in the graph. That's how I understood the concept, especially in reference to lazy-analysis, on-open analysis, and looking at ns-forms. 

However, I now believe you are saying shallow-analyze all internal files, and from that build a series of dependency graphs of dependants.
```clojure
[lib.core [example.impl-test example.impl example.extension]]
[example.impl [example.main]]
[example.extension [example.main]]
 ````
Correct? 

This would leave out `user` because of shallowness and kondo not seeing the unrequired lib (which I'll argue should be analyzed as a usage but still linted).

It's not uncommon to need to recurse into arbitrary forms in the main program either, look at the [sente docs](https://github.com/ptaoussanis/sente/blame/7a1fad84cad9839834648a3d89dce8c1807d2f70/README.md#L90-L99) 

I've seen macros do things like:
```clojure
(defmacro with-my-assoc [& body]
  `(do
     (require (quote [~'medley.core :refer [~'assoc-some]]))
     ~@body))

(defn x []
  (with-my-assoc
    (assoc-some {} :x nil)))
```

External files also do these things pretty often, and if you jump into a jar it's pretty frustrating if you can't navigate the code. Yet, it's less critical to have a complete view of external code as you aren't (hopefully) refactoring it, so if some usages are missed maybe that's not the end of the world.

> clj-kondo knows of a few special cases that are common within the community, but if you go outside that realm, we won't help you. In particular, we cannot guarantee that we "operated on a complete view of the project".

> (If you call import-vars inside of another function, that's much like calling def within a function, which [is discouraged](https://guide.clojure.style/#dont-def-vars-inside-fns). I'm not saying that people don't do it, but that I don't expect clj-kondo or clojure-lsp to ever automatically handle cases like that. For example, should clj-kondo resolve vars in that namespace? What if that function is never called?)

Sure, static analysis will never pick up dynamic calls like `(require [(symbol user-ns) :as (symbol (str "prefix-" user-alias))])` but kondo hooks have a lot of power to pick that up if it's done in a macro. Kondo picks up nested defs and requires now.

My contention is that if kondo can (or should) pick up a dependency through full-analysis, but will not pick it up in shallow analysis then it would be clojure-lsp's fault if we used the shallow graph.  If I build dep-graphs from full-analysis and shallow-analysis many of the projects I've encountered would differ significantly.

**Comment by mainej** [comment]
Cool, I think we're getting closer to a shared understanding. And I'm starting to see where I haven't explained things well.

First let me acknowledge several things I think we agree on, and then I'll follow up with more questions about shallow analysis.

First we agree (I think) that a dep-graph could be useful for queries. You're right—that's what I'm interested in building now. And we agree (I think) that for the queries to be correct, we need a **complete** dep graph. (Your second diagram is close to what I had in mind, where the graph includes every file, not just files that have been visited and their neighbors. My one quibble with your diagram is that it looks more uni-directional than what I'm envisioning. I'd like the dep graph to be bidirectional, to be able to easily traverse from either direction: from a namespace to its dependencies (the namespaces it uses), but also from a namespace to its dependents (the namespaces that use it.) Maybe that's what you meant—it's hard to draw graphs in ASCII.)

We also agree that lazy analysis breaks several major features. So far we've catalogued problems with full-project diagnostics, workspace symbols, parts of completion, and keyword usages. There may be other problems, but the existing list is already pretty high-profile. All that makes lazy analysis less appealing. I'm not planning to work on it.

And finally, we even agree that shallow analysis isn't the thing to start with, though for different reasons. My reason is that it would leave holes in other features (most notably keyword usages and full project analysis.) Your reason is primarily that you think it doesn't gather enough information to build a correct dep graph.

Does that sound right so far?

If so, then I concur—let's not change the analysis pipeline right now, not even to add shallow analysis of project files.

However, in case I ever come back to the idea of shallow analysis, I'd still like to understand why you feel it wouldn't build a complete dep graph.

To make the discussion more concrete, I'd like to ask: does clj-kondo ever create a `:namespace-usage` element because of a `require` inside of a `def`? That's the key question for whether shallow analysis could make a complete dep graph.

You provided several pieces of evidence that a `require` inside a `def` _does_ create a namespace-usage. But I'm not convinced of that. For me, when I try any of your examples I get an unresolved-var or unresolved-namespace lint, suggesting that a namespace-usage _isn't_ created. So let's go over your examples in more detail, to see what I'm missing.

Let's start with the code I provided. I'm also on master, but here's what I see:

<img width="601" alt="Screen Shot 2022-05-27 at 11 25 16 PM" src="https://user-images.githubusercontent.com/6774/170816232-8876a027-acf4-4977-a4ab-ba12bbed67ed.png">

Why are we seeing different things? Why do I see an unresolved-namespace lint but you don't?

One possibility is that the file you were in has some other `require` of medley. If it does, then I believe clj-kondo is able to resolve `map-vals` inside the `defn` for you not because of the `require` that's in the `defn`, but because of that other `require`.

For example, for me, the lint goes away if I move the snippet below a comment block which requires medley:

<img width="274" alt="Screen Shot 2022-05-28 at 12 55 23 AM" src="https://user-images.githubusercontent.com/6774/170816394-ff99866d-4055-42b2-b9ce-8dbcb1d1a24f.png">

It also goes away if I require medley in the `ns` declaration or at the top level.

If my example truly works for you in an empty file, why? What's different about our config?

Moving on to your other examples... 

> the [sente docs](https://github.com/ptaoussanis/sente/blame/7a1fad84cad9839834648a3d89dce8c1807d2f70/README.md#L90-L99)

Again, shallow analysis doesn't stop at the top level, it stops before analyzing the children of a def, so shallow analysis would pick up all those sente defs.

You also gave an example of a macro `with-my-assoc`. A macro like that is definitely possible, but my question is... what does the hook look like that fixes the lint? Does that hook make clj-kondo create a namespace-usage whenever it sees that macro? I'm certainly not an expert at hooks, but I don't see how it's possible to write such a hook. As far as I'm aware, the best a hook can do is rewrite your sample code into this:

```clojure
(defn x []
  (require '[medley.core :refer [assoc-some]])
  (assoc-some {} :x nil))
```

Or possibly this:

```clojure
(defn x []
  (medley.core/assoc-some {} :x nil))
```

But if I make those conversions manually, for me, neither one fixes the lint. Am I wrong about this? It's quite possible I don't know enough about hooks, so I'd love to understand if there's a way to make a hook that would let clj-kondo analyze this as a namespace-usage.

And finally, `user`.

```clojure
(ns user)
(comment
  (lib.core/foo ...) ;; called from nrepl that has ns loaded through `example.main`
  )
```

For me, currently clj-kondo doesn't resolve this as a call to `lib.core/foo`, so it doesn't show up as a usage. Should clj-kondo analyze it differently? That is, is this a missing analyzer feature? Maybe. But if that feature were added, then I think it'd be reasonable to add a `:namespace-usage` at the same time. And since none of this `user` code is inside a `def`, it'd all get picked up by shallow analysis.

Anyway, from what I'm seeing shallow analysis would pick up just as many namespace-usages as full analysis. But, I'm still curious if there are counterexamples.

**Comment by ericdallo** [comment]
> A big thing that I think we've gone away from is picking up changes from the filesystem. It is very important that if you git pull new files that usages and lints are updated in the editor.

This is something the LSP protocol already has a canonical way to solve this issue, the server should listen to `didChangeWatchedFiles` that clients should send (and most/all of them already send) but clojure-lsp ignores that, I tried in the past to process the changes but that caused lots of performance issues, I think we should try again as we changed a lot the way we analyze files, so to me the `git pull`  would in the end just trigger multiple changes similar to didChange

**Comment by snoe** [comment]
> Does that sound right so far?

Yup. I think we got it, thanks for your patience.

> Let's start with the code I provided. I'm also on master, but here's what I see:

Hrm, I switched laptops over the weekend, and now I'm also seeing what you see. Maybe I did something wrong like you guessed.

> Anyway, from what I'm seeing shallow analysis would pick up just as many namespace-usages as full analysis. But, I'm still curious if there are counterexamples.

So, I suppose there's two parts here.
1. This can be tested when this is revisited, run shallow and full against a few tricky projects and see if the dep graphs match. If we added that to CI we could follow improvements to kondo. 
2. I want to consider what should be analyzed rather than what is currently analyzed. I don't want to box ourselves in if analyzers pick up a static `require` in an unexpected place in the future.

**Comment by borkdude** [comment]
I appreciate all the effort! I think what would be good is a _problem statement_ first.

I haven't heard many complaints about clojure-lsp and initial indexing time. I did hear several complaints about stability (editor stuck or slow due to loops). 

Introducing big changes like this seem risky when it comes to stability. I wouldn't want to risk users to go away when the problem being solved wasn't a problem for them in the first place?

**Comment by mainej** [comment]
@borkdude that's a reasonable suggestion.

I'll divide this into two problem statements.

## Problem 1: Startup is slow on very large projects

Here are (edited) logs for an uncached startup on my machine of https://github.com/metabase/metabase, a large project (4MB in src).

```
2022-06-06T16:01:24.229Z  ERROR [clojure-lsp.db:79] - [DB] No cache DB file found
2022-06-06T16:01:24.325Z  INFO [clojure-lsp.classpath:54] - Finding classpath
2022-06-06T16:02:30.746Z  INFO [clojure-lsp.crawler:58] - External classpath paths analyzed, took 22455ms
2022-06-06T16:02:38.126Z  INFO [clojure-lsp.crawler:191] - [Startup] Analyzing source paths for project root /Users/jmaine/workspace/opensource/metabase
2022-06-06T16:03:34.540Z  INFO [clojure-lsp.crawler:27] - [Startup] Project only paths analyzed, took 56411ms
2022-06-06T16:03:35.281Z  INFO [clojure-lsp.handlers:93] - :initialize 131065ms
2022-06-06T16:03:39.985Z  INFO [lsp4clj.core:333] - [Server] Initialized!
2022-06-06T16:05:41.207Z  INFO [clojure-lsp.server:82] - Publishing 2 diagnostics
```

It takes about 2 minutes for the clojure-lsp server to become interactive. About 2/3 of that time is analyzing (22 seconds external, 56 seconds project). During this time the CPUs are fully utilized.

(After the server is started, it spends another 2 minutes publishing all the diagnostics, which seems surprisingly long. During this time the CPUs are half utilized. Something to research.)

This first startup caches its analysis of the external files, so subsequent startups are faster.

```
2022-06-06T17:47:28.898Z  INFO [clojure-lsp.db:72] - [DB] Reading transit analysis cache from /Users/jmaine/workspace/opensource/metabase/.lsp/.cache/db.transit.json db took 1207ms
2022-06-06T17:47:36.507Z  INFO [clojure-lsp.crawler:160] - [Startup] Using cached db for project root /Users/jmaine/workspace/opensource/metabase
2022-06-06T17:47:38.575Z  INFO [clojure-lsp.crawler:191] - [Startup] Analyzing source paths for project root /Users/jmaine/workspace/opensource/metabase
2022-06-06T17:48:21.720Z  INFO [clojure-lsp.crawler:27] - [Startup] Project only paths analyzed, took 43142ms
2022-06-06T17:48:22.142Z  INFO [clojure-lsp.handlers:93] - :initialize 54471ms
2022-06-06T17:48:24.125Z  INFO [lsp4clj.core:333] - [Server] Initialized!
2022-06-06T17:50:29.705Z  INFO [clojure-lsp.server:82] - Publishing 2 diagnostics
```

It takes slightly less than 1 minute to become interactive, about 3/4 analyzing. (And again, there's an additional 2 minutes publishing diagnostics.)

@ericdallo has said he'd like to make startup faster for two reasons. First, the uncached startup time is a problem if it affects the "marketing" of clojure-lsp. It can lead to a perception that clojure-lsp is slow. This problem is most likely to affect new users, who are forming their first impressions. Second, the cached startup time affects day-to-day usage impacting how quickly one can get to work after re-opening the project or re-starting clojure-lsp.

### Alternatives

1. It's fine as is. Users are willing to wait.
2. Reduce and/or postpone analysis. See the rest of this Issue for details.
3. Cache project analysis.

The first—and default—option is to do nothing. Users can live with the startup time. We may see startup time improvements from other performance work, but the best case scenario is that 100% of startup is time analyzing. For metabase that would mean startup times of 80s uncached and 43s cached. This approach is reasonable. Not many users work in truly massive projects. We haven't had complaints recently from other users.

The second option is to try to reduce analysis during startup, or postpone it until later. The rest of this proposal described one possible way to do this, using shallow and lazy analysis respectively to reduce and postpone analysis. We've done a pretty deep dive into the pros and cons of this approach. The biggest con is that it leaves holes in the analysis which would affect a variety of features. Though it may be possible to patch those holes with additional engineering effort in some cases, for now we've decided to abandon this path. 

A third option—which has been discussed elsewhere, but to my knowledge never formalized—is to cache project analysis. We already cache external analysis, but this cache is easier to maintain because external dependencies rarely change, so they can be cached at startup. Project files, however, are continually changing, which would make cache maintenance more dynamic. Further, the cache key for the external files is based purely on file timestamps. But for project files, clojure-lsp's analysis can match the file on disk, be for an unsaved version of the file in an open buffer, or not match at all, perhaps because of a source-control change. And finally, there is a small concern about the size of the cache file. For metabase the cache file is currently 41MB. Adding project analysis would make the file larger, and may mean regularly reading and writing the file.

### Plan

Option 1: do nothing.

## Problem 2: It is easy to introduce performance problems, and performance related improvements are challenging

At the root of several recent performance problems, there's a common cause: scanning too much of the analysis, too often. As references I've pulled up some PRs, though just ones that I've worked on:

https://github.com/clojure-lsp/clojure-lsp/pull/971
https://github.com/clojure-lsp/clojure-lsp/pull/934
https://github.com/clojure-lsp/clojure-lsp/pull/933
https://github.com/clojure-lsp/clojure-lsp/pull/892
https://github.com/clojure-lsp/clojure-lsp/pull/885
https://github.com/clojure-lsp/clojure-lsp/pull/861
https://github.com/clojure-lsp/clojure-lsp/pull/771
https://github.com/clojure-lsp/clojure-lsp/pull/764

These problems often manifest when the straight-forward algorithm gets quadratically worse in large projects. As evidenced by the above PRs, it's possible to fix these problems, but the fixes currently come at the cost of more complex algorithms and corresponding complex code.

In my judgement, this is largely because clojure-lsp lacks tools to efficiently narrow the analysis. Currently, clojure-lsp developers have three ways to narrow the analysis:

1. Filter analysis for a file
2. Filter analysis for all the project files
3. Do not filter the analysis

In metabase there are 551K elements in the full analysis, and project-only analysis is not much better, at 416K elements. Even a single scan over this many elements, without any transformations, takes over 40ms.

Because every query has to scan over so many elements, every feature has to be careful to minimize the number of times it scans the analysis. If any feature ever does a nested scan, it is guaranteed to lock up the CPU.

### Proposal

As described above, many queries could narrow to just a few files worth of analysis, if they had access to a graph of the dependencies between namespaces and metadata about the files that define those namespaces. In these cases we would decouple the performance of the queries from the size of the project, meaning those queries would stay efficient regardless of how big a project grew.

This is not a panacea—in some cases we will still have to use the full analysis, and in all cases we should still strive for efficient algorithms.

But even algorithms that already do only a few scans will benefit, if we can go from scanning hundreds of thousands of elements to a few thousand. And as new feature are introduced, in simple cases it will be easier to write efficient algorithms, and in more complex cases easier to refactor when performance problems appear.

### Plan

Pursue development of a dependency graph. I have this work in progress. I've been busy with other things, so I'm moving a bit slow, but the results so far are promising. I'll present the work in a PR later.

**Comment by mainej** [comment]
I think everyone in this discussion is aware of this, but for completeness I should clarify the proposal in the second problem statement. I'm going to edit the original problem statement comment. I'll repeat the edits below.

Generally queries are all about filtering from a large number of analysis elements to a small number of "interesting" elements. In clojure-lsp, the analysis elements are indexed by filename. So, the first step in filtering is always to figure out which files to look at. 

With our existing tools, we can efficiently find the analysis elements for a single file. We also have a way to filter to project-internal files. Otherwise we have to use all the elements.

Having a dep graph would help, for example in go-to definition. Currently, if the cursor is on a var usage `some-ns/some-fn|`, we know the name of the file where the var is used, but not the names of the files which define `some-ns`. So we can quickly locate the var usage by looking at its file's analysis, but to find the var definition we have no option but to scan every element in the analysis. If we had a dep-graph, we could look up the files which define `some-ns`, and filter to their elements.

Similarly in completion, if a user has typed `some-ns/some-f|` to make completion suggestions, we need to look for var definitions only in files which define `some-ns`, not every file.

**Comment by ericdallo** [comment]
## About problem 1

> It takes slightly less than 1 minute to become interactive, about 3/4 analyzing. (And again, there's an additional 2 minutes publishing diagnostics.)

This was actually on purpose, we were receiving reports of high CPU after project started, I changed to use a less parallel approach similar to how kondo does, and I'm liking the results so far, it's not that fast, but the CPU usage is way better.

Your explanation is correct and I agree with the alternatives, I'd like just to add a little bit about why I like options 1 and 3.
With important improvements since the start of this year, as you said, we didn't receive any project performance reports, the improvements on the external cache analysis and CPU usage really helped with that, so option 1 is really something to consider because besides those huge project (I mean as big as metabase, which even on Nubank I know only 1 or 2 of that size), clojure-lsp is performing really fine on the rest IMHO, and with the external analysis cache improvements, my only concern now is project size, not dependencies anymore, which is why metabase takes that time.

Even so, there are "quick-wins" that could improve the performance for metabase-like projects, like the project analysis cache, if that works as well external cache worked, the only issue would be opening the first time the project, which would take a couple of minutes on the worst scenario and I think that will be fine for most people.

## About problem 2

I think this is a more important problem than 1, because it affects the whole clojure-lsp proccess/developer session of that project, not only the startup, and if we think in a feature-flag way to testing that dependency graph, sounds like a good idea to me.
